src/reduce.js

```diff
@@ -166,13 +166,14 @@ export function reduceL1(input, options = {}) {
 
 /* The l2 reducer */
 export function l2Reducer(previousValue, currentValue, currentIndex, array) {
+  let sumOfSquares;
   if (currentIndex == 1) {
-    const sumOfSquares = previousValue * previousValue + currentValue * currentValue;
-    return sumOfSquares;
+    sumOfSquares = previousValue * previousValue + currentValue * currentValue;
   } else {
-    const sumOfSquares = previousValue + currentValue * currentValue;
-    return (currentIndex === array.length - 1) ? Math.sqrt(sumOfSquares) :sumOfSquares;
+    sumOfSquares = previousValue + currentValue * currentValue;
   }
+
+  return (currentIndex === array.length - 1) ? Math.sqrt(sumOfSquares) : sumOfSquares;
```

Comment:
Hmm, it's kinda complicated to figure this out, given the special cases for the first and last element. Looking at what you did for other similar cases like how you implemented `reduceSumSquare` and `reduceLogSum`, would it be simpler and more compositional to implement 2D Lebesgue pooling (`y = (x1^p + x2^p + ... + xn^p) ^ (1/p)`) as:

```
export function l2Pool2d(input, options = {}) {
  return pow(pool2d(pow(input, new Scalar(2)), sumReducer, options), new Scalar(0.5));
}
```

```js
export function reduceSumSquare(input, options = {}) {
  return reduceSum(pow(input, new Scalar(2)), options);
}

export function reduceLogSum(input, options = {}) {
  return log(reduceSum(input, options));
}
```

Reply:
Updated! Please take another look, thanks!